prow/githuboauth/githuboauth.go

```diff
@@ -144,6 +179,10 @@ func (ga *Agent) HandleLogout(client OAuthClient) http.HandlerFunc {
 // the final destination in the config, which should be the front-end.
 func (ga *Agent) HandleRedirect(client OAuthClient, getter GitHubClientGetter) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
+		finalRedirectURL, err := r.URL.Parse(r.URL.Query().Get("dest"))
```

Comment:
This is broadly considered to be unsafe: by accepting any arbitrary URL as a redirect destination, we are acting as an "open redirect" - that is, an attacker can make a prow.k8s.io link go to any website (in this case, with a GitHub login in the middle), which opens a nominal phishing vector.

These redirects should be locked down more - IIRC an earlier version of this PR only let you access pages on the same domain, which largely mitigates the issue. An alternative approach is an explicit whitelist of valid redirect targets.

Reply:
I believe this does the same thing as the earlier version, just using the url library. `r.URL.Parse` uses the hostname from the request URL, which I think has to be prow.k8s.io?

Reply:
`(*URL).Parse()` parses a string in the context of the original URL. Given a string that contains only a path, like `/pr`, `r.URL.Parse("/pr")` will use the initial context and produces a URL like `https://prow.k8s.io/pr`.

However, given a complete URL, e.g. `https://evil.com/foo`, the context is no longer necessary, and so `r.URL.Parse("https://evil.com/foo")` will produce `https://evil.com/foo` ([try it out!](https://play.golang.org/p/Tpm0iKB1gvK)).

The previous string concatenation approach was safer: there is no way (that I am aware of) to get out of prow.k8s.io like that.

Either reverting to string concatenation or verifying that `finalRedirectURL` has the host you expect should fix the problem.

Reply:
I check now that there's no host specified in `finalRedirectURL`, which should make sure it's redirected back to the same host.